src/realm/sync/socket_provider.hpp

```diff
@@ -0,0 +1,229 @@
+///////////////////////////////////////////////////////////////////////////////
+//
+// Copyright 2022 Realm Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+// http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+//
+///////////////////////////////////////////////////////////////////////////////
+
+#pragma once
+
+#include <map>
+#include <memory>
+#include <string>
+
+#include <realm/status.hpp>
+
+#include <realm/sync/config.hpp>
+
+#include <realm/util/functional.hpp>
+#include <realm/util/optional.hpp>
+#include <realm/util/span.hpp>
+
+namespace realm::sync {
+
+struct WebSocketEndpoint;
+struct WebSocketInterface;
+struct WebSocketObserver;
+
+/// Sync Socket interface that provides the event loop and WebSocket factory
+/// used by the SyncClient.
+///
+/// All callback and event operations in the SyncClient must be completed in
+/// the order in which they were issued (via post() or timer) to the event
+/// loop and cannot be run in parallel. It is up to the custom event loop
+/// implementation to determine if these are run on the same thread or a
+/// thread pool as long as it is guaranteed that the callback handler
+/// functions are processed in order and not run concurrently.
+///
+/// The implementation of a SyncSocketInterface must support the following
```

Comment:
you may want to update the comments (eg, SyncSocketInterface -> SyncSocketProvider)